test/new-e2e/tests/agent-shared-components/forwarder/nss_failover_test.go

```diff
@@ -0,0 +1,302 @@
+func setHostEntry(t *testing.T, vm client.VM, hostname string, hostIP string) {
+	// we could remove the line and then add the new one,
+	// but it's better to avoid not having the line in the file between the two operations
+
+	t.Logf("set host entry for %s: %s", hostname, hostIP)
+
+	hostfile := vm.Execute("sudo cat /etc/hosts")
+
+	// pattern to match the hostname entry
+	hostPattern := fmt.Sprintf("^.* %s$", regexp.QuoteMeta(hostname))
+	// enable multi-line mode in the Go regex
+	goHostPattern := fmt.Sprintf("(?m:%s)", hostPattern)
+	matched, err := regexp.MatchString(goHostPattern, hostfile)
+	require.NoError(t, err)
+
+	entry := fmt.Sprintf("%s %s", hostIP, hostname)
+	if matched {
+		t.Logf("replace existing host entry for %s (%s)", hostname, hostIP)
+		vm.Execute(fmt.Sprintf("sudo sed -i 's/%s/%s/g' /etc/hosts", hostPattern, entry))
+	} else {
+		t.Logf("append new host entry for %s (%s)", hostname, hostIP)
+		vm.Execute(fmt.Sprintf("echo '%s' | sudo tee -a /etc/hosts", entry))
+	}
+}
+
+// enforceNSSwitchFiles ensures /etc/nsswitch.conf uses `files` first for the `hosts` entry
+// so that an NSS query uses /etc/hosts before DNS
+func enforceNSSwitchFiles(t *testing.T, vm client.VM) {
+	// for the specifics of the nsswitch.conf file format, see its man page
+	//
+	// in short, the hosts line starts with "hosts:", then a whitespace separated list of "services"
+	// each service can be followed by actions in the format [STATUS=ACTION] or [!STATUS=ACTION]
+	// we want to have the "files" service first without any action after
+
+	t.Logf("enforce using files first in NSS")
+
+	nsswitchfile := vm.Execute("sudo cat /etc/nsswitch.conf")
+
+	// enable multi-line mode in the Go regex
+	regex, err := regexp.Compile(`(?m:^hosts:\s+(.*)$)`)
+	require.NoError(t, err)
+
+	if regex.MatchString(nsswitchfile) {
+		matches := regex.FindStringSubmatch(nsswitchfile)
+		require.NotNil(t, matches)
+
+		services := strings.Fields(matches[1])
+		if len(services) == 0 || services[0] != "files" || (len(services) >= 2 && services[1][0] == '[') {
+			t.Logf("replace existing hosts entry in /etc/nsswitch.conf")
+			// add `files` before previous services
+			vm.Execute(`sudo sed -E -i 's/^hosts:(\s+)(.*)$/hosts:\1files \2/g' /etc/nsswitch.conf`)
+		}
+	} else {
+		t.Logf("add hosts entry in /etc/nsswitch.conf")
+		vm.Execute("echo 'hosts: files' | sudo tee -a /etc/nsswitch.conf")
+	}
+}
+
+// hostIPFromURL extracts the host from the given URL and returns any IP associated to that host
+// or an error
+func hostIPFromURL(fakeintakeURL string) (string, error) {
+	parsed, err := url.Parse(fakeintakeURL)
+	if err != nil {
+		return "", err
+	}
+
+	host := parsed.Hostname()
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return "", err
+	}
+
+	if len(ips) == 0 {
+		return "", fmt.Errorf("no ip for host %s", host)
+	}
+
+	// return any valid ip
+	return ips[0].String(), nil
+}
+
+func getAgentConfig(interval int) string {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
I see this is used only once with a constant value. What about moving it to a constant embedded file instead ? Do you plan to change the interval ?

Reply:
I initially had more things in that config but was able to factor them out
My main reason for having those setting in the Go code is that I use the value of the interval in the test, and I wanted to avoid having it hardcoded in different files.
I guess I could use an embedded template file though, that would be cleaner

Reply:
Used embedded template files :+1: